devices/src/virtio/vsock/unix/muxer_rxq.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// `MuxerRxQ` implements a helper object that `VsockMuxer` can use for queuing RX (host -> guest)
+/// packets (or rather instructions on how to build said packets).
+///
+/// Under ideal operation, every connection, that has pending RX data, will be present in the muxer RX
+/// queue. However, since the RX queue is smaller than the connection pool, it may, under some conditions,
+/// become full, meaning that it can no longer account for all the connections that can yield RX data.
+/// When that happens, we say that it is no longer "synchronized" (i.e. with the connection pool).
+/// A desynchronized RX queue still holds valid data, and the muxer will continue to pop packets from
+/// it. However, when a desynchronized queue is drained, additional data may still be available, so the
+/// muxer will have to perform a more costly walk of the entire connection pool to find it.
+/// This walk is performed here, as part of building an RX queue from the connection pool. When an
+/// out-of-sync is drained, the muxer will discard it, and attempt to rebuild a synced one.
+///
+use std::collections::{HashMap, VecDeque};
+
+use super::super::VsockChannel;
+use super::defs;
+use super::muxer::{ConnMapKey, MuxerRx};
+use super::MuxerConnection;
+use super::{Error, Result};
+
+/// The muxer RX queue.
+///
+pub struct MuxerRxQ {
+    /// The RX queue data.
+    q: VecDeque<MuxerRx>,
+    /// The RX queue sync status.
+    synced: bool,
+}
+
+impl MuxerRxQ {
+    const SIZE: usize = defs::MUXER_RXQ_SIZE;
+
+    /// Trivial RX queue constructor.
+    ///
+    pub fn new() -> Self {
+        Self {
+            q: VecDeque::with_capacity(Self::SIZE),
+            synced: true,
+        }
+    }
+
+    /// Attempt to build an RX queue, that is synchronized to the connection pool.
+    /// Note: the resulting queue may still be desynchronized, if there are too many connections
+    ///       that have pending RX data. In that case, the muxer will first drain this queue, and
+    ///       then try again to build a synchronized one.
+    ///
+    pub fn from_conn_map(conn_map: &HashMap<ConnMapKey, MuxerConnection>) -> Self {
```

Comment:
That was my initial approach, until I noticed that the queue basically gets drained and re-created, for all intents and purposes, every time it gets desynchronized. Actually, that's where the sync/desync terminology came from. In retrospect, now, that it gets destroyed/re-created, something like "degraded" instead of "desynchronized" seems more appropriate.